pkg/lib/cockpit-upload-helper.ts

```diff
@@ -0,0 +1,142 @@
+
+    // This variable is the most important thing in this function.  The main
+    // upload loop will do work for as long as it can, and then it .block()s on
+    // the waiter until something changes (ack, close, abort, etc).  All of
+    // those things call .unblock() to resume the loop.
+    const event_waiter = new Waiter();
+
+    if (signal) {
+        signal.throwIfAborted(); // early exit
+        signal.addEventListener('abort', () => { event_waiter.unblock() });
+    }
+
+    debug('requesting channel');
+    const channel = cockpit.channel({
+        payload: 'fsreplace1',
+        path: destination,
+        binary: true,
+        'send-acks': 'bytes',
+        ...options,
+    });
+    channel.addEventListener('control', (_ev, message) => {
+        debug('control', message);
+        if (message.command === 'ack') {
+            cockpit.assert(typeof message.bytes === 'number', 'bytes not a number');
+            delivered += message.bytes;
+            if (progress) {
+                debug('progress', delivered);
+                progress(delivered);
+            }
+            outstanding -= message.bytes;
+            debug('outstanding -- to', outstanding);
+            event_waiter.unblock();
+        }
+    });
+    channel.addEventListener('close', (_ev, message) => {
+        debug('close', message);
+        close_message = message;
+        event_waiter.unblock();
+    });
+
+    try {
+        debug('starting file send');
+        const reader = stream.getReader({ mode: 'byob' }); // byob to choose the block size
+        let eof = false;
+
+        // eslint-disable-next-line no-unmodified-loop-condition
+        while (!close_message) {
+            /* We do the following steps for as long as the channel is open:
+             *  - if there is room to write more data, do that
+             *  - otherwise, block on the waiter until something changes
+             *  - in any case, check for cancellation, repeat
+             * The idea here is that each loop iteration will `await` one
+             * thing, and once it returns, we need to re-evaluate our state.
+             */
+            if (!eof && outstanding < FLOW_WINDOW) {
+                const { done, value } = await reader.read(new Uint8Array(BLOCK_SIZE));
+                if (done) {
+                    debug('sending done');
+                    channel.control({ command: 'done' });
+                    eof = true;
+                } else {
+                    debug('sending', value.length, 'bytes');
+                    channel.send(value);
+                    outstanding += value.length;
+                    debug('outstanding ++ to', outstanding);
+                }
+                if (signal) {
+                    signal.throwIfAborted();
+                }
+            } else {
+                debug('sleeping', outstanding, 'of', FLOW_WINDOW, 'eof', eof);
+                await event_waiter.block();
```

Comment:
I had to think a bit if this would not wait on an `ack` but yeah it can't!

Reply:
How do you mean?

Reply:
Resolved in matrix.